src/wallet/spend.cpp

```diff
@@ -575,11 +575,12 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& a
     // Deduct preset inputs amount from the search target
     CAmount selection_target = nTargetValue - pre_set_inputs.total_amount;
 
-    // If automatic coin selection was disabled, we just want to return the preset inputs result
-    if (!coin_control.m_allow_other_inputs) {
+    // If automatic coin selection was disabled or if we already covered the
+    // target with the preset inputs: try to return the result right away
+    if (!coin_control.m_allow_other_inputs || selection_target <= 0) {
         // 'selection_target' is computed on `PreSelectedInputs::Insert` which decides whether to use the effective value
         // or the raw output value based on the 'subtract_fee_outputs' flag.
-        if (selection_target > 0) return std::nullopt;
+        if (selection_target > 0) return std::nullopt; // fails only if 'allow_other_inputs=false'
         SelectionResult result(nTargetValue, SelectionAlgorithm::MANUAL);
         result.AddInputs(pre_set_inputs.coins, coin_selection_params.m_subtract_fee_outputs);
         result.ComputeAndSetWaste(coin_selection_params.min_viable_change, coin_selection_params.m_cost_of_change, coin_selection_params.m_change_fee);
```

Comment:
in d54606dad097ac050dfdd1c820e3d7172b036146
IMO this boolean logic:
```cpp
    // If automatic coin selection was disabled or if we already covered the
    // target with the preset inputs: try to return the result right away
    if (!coin_control.m_allow_other_inputs && selection_target > 0) return std::nullopt;

    if (selection_target <= 0) {
        SelectionResult result(nTargetValue, SelectionAlgorithm::MANUAL);
        result.AddInputs(pre_set_inputs.coins, coin_selection_params.m_subtract_fee_outputs);
        result.ComputeAndSetWaste(coin_selection_params.min_viable_change, coin_selection_params.m_cost_of_change, coin_selection_params.m_change_fee);
        return result;
    }
```
is easier to understand than this one:
```cpp
    // If automatic coin selection was disabled or if we already covered the
    // target with the preset inputs: try to return the result right away
    if (!coin_control.m_allow_other_inputs || selection_target <= 0) {
        // 'selection_target' is computed on `PreSelectedInputs::Insert` which decides whether to use the effective value
        // or the raw output value based on the 'subtract_fee_outputs' flag.
        if (selection_target > 0) return std::nullopt; // fails only if 'allow_other_inputs=false'
        SelectionResult result(nTargetValue, SelectionAlgorithm::MANUAL);
        result.AddInputs(pre_set_inputs.coins, coin_selection_params.m_subtract_fee_outputs);
        result.ComputeAndSetWaste(coin_selection_params.min_viable_change, coin_selection_params.m_cost_of_change, coin_selection_params.m_change_fee);
        return result;
    }
```

Reply:
yeah, better.